torch/csrc/distributed/c10d/default_comm_hooks.cpp

```diff
@@ -16,21 +16,23 @@ c10::intrusive_ptr<c10::ivalue::Future> AllReduceCommHook::runHook(
 
 c10::intrusive_ptr<c10::ivalue::Future> FP16CompressCommHook::runHook(
     GradBucket& bucket) {
-  auto& tensor = bucket.getBufferRef();
-  tensor.copy_(tensor.to(torch::kFloat16));
-  std::vector<at::Tensor> tensors = {tensor};
+
+  auto compressed_tensor = bucket.getBufferRef().to(torch::kFloat16);
   // Apply the division first to avoid overflow.
-  tensors[0] /= state_->getSize();
+  compressed_tensor /= state_->getSize();
+  std::vector<at::Tensor> tensors = {compressed_tensor};
 
   auto allreduce_fut = state_->allreduce(tensors)->getFuture();
-  auto decompress = [](c10::ivalue::Future& allreduce_fut) {
+  auto decompress = [bucket](c10::ivalue::Future& allreduce_fut) {
```

Comment:
bucket -> &bucket 
This can avoid copying vector data fields of `GradBucket`, such as `offsets`, `lengths`, `sizes_vec`. Copying these vectors can be non-trivial if there are a large number of parameters.